python/samples/concepts/auto_function_calling/nexus_raven.py

```diff
@@ -0,0 +1,361 @@
+# Copyright (c) Microsoft. All rights reserved.
+
+import ast
+import asyncio
+import json
+import math
+from collections.abc import AsyncGenerator
+from html import escape
+from typing import Annotated, Any, Literal
+
+from huggingface_hub import AsyncInferenceClient
+from pydantic import HttpUrl
+
+from semantic_kernel import Kernel
+from semantic_kernel.connectors.ai.chat_completion_client_base import ChatCompletionClientBase
+from semantic_kernel.connectors.ai.open_ai import (
+    OpenAIChatCompletion,
+)
+from semantic_kernel.connectors.ai.open_ai.prompt_execution_settings.open_ai_prompt_execution_settings import (
+    OpenAIChatPromptExecutionSettings,
+)
+from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
+from semantic_kernel.connectors.ai.text_completion_client_base import TextCompletionClientBase
+from semantic_kernel.contents import (
+    ChatHistory,
+    ChatMessageContent,
+    FunctionCallContent,
+    StreamingChatMessageContent,
+    TextContent,
+)
+from semantic_kernel.contents.function_result_content import FunctionResultContent
+from semantic_kernel.functions import KernelArguments, kernel_function
+
+kernel = Kernel()
+
+#########################################################################
+# Step 0: Define a custom AI Service, with Prompt Execution settings. ###
+# This uses huggingface_hub package, so install that if needed.       ###
+#########################################################################
+
+
+class NexusRavenPromptExecutionSettings(PromptExecutionSettings):
+    do_sample: bool = True
+    max_new_tokens: int | None = None
+    stop_sequences: Any = None
+    temperature: float | None = None
+    top_p: float | None = None
+
+    def prepare_settings_dict(self, **kwargs) -> dict[str, Any]:
+        """Prepare the settings dictionary."""
+        return self.model_dump(
+            include={"max_new_tokens", "temperature", "top_p", "do_sample", "stop_sequences"},
+            exclude_unset=False,
+            exclude_none=True,
+            by_alias=True,
+        )
+
+
+class NexusRavenCompletion(TextCompletionClientBase, ChatCompletionClientBase):
+    """To use this class, you should have installed the ``huggingface_hub`` package, and
+    the environment variable ``HUGGINGFACEHUB_API_TOKEN`` set with your API token,
+    or given as a named parameter to the constructor."""
+
+    client: AsyncInferenceClient
```

Comment:
Any chance this will become a hugging face connector?

Reply:
Not in it's current form, I was thinking we should have a look to see the best approach for huggingface_hub models, whether as a subtype to the current one or as a seperate one, this one is highly specialized for this scenario, that is why I named it NexusRavenCompletion and not something else!